enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1032 @@
+
+The procedure to resize a Control Plane today is lengthy and very involved. It takes a significant amount of time for
+an OpenShift expert to perform. We also document this process for our end users, however due to the complexity of the
+procedure, there is discomfort in recommending this procedure to end users who may not be as familiar with the product.
+
+To ensure the long term usability of OpenShift, we must provide an automated way for users to scale their Control Plane
+as and when their cluster needs additional capacity.
+
+As HyperShift adoptions grows, HyperShift will solve the same issue by running hosted Control Planes within management
+clusters. However, HyperShift is not a suitable product for all OpenShift requirements (for example the first management
+cluster) and as such, we see that HA clusters will continue to be used and we must solve this problem to allow the
+continued adoption of HA clusters.
+
+### Goals
+
+* Provide a "best practices" approach to declarative management of Control Plane Machines
+* Allow users to "1-click" scale their Control Plane to large or smaller instances
+* Allow the adoption of existing Control Planes into this new mechanism
+* Provide a safe mechanism to perform the sensitive operation of scaling the Control Plane and provide adequate
+  feedback to end users about progress of the operation
+* Allow users to opt-out of Control Plane management should our mechanism not fit their needs
+* Allow users to customise rolling update strategies based on the needs of their environments
+
+### Non-Goals
+
+* Allow horizontal scaling of the Control Plane (this may be required in the future, but today is considered
+  unsupported by OpenShift)
+* Management of the etcd cluster state (the etcd operator will handle this separately)
+* Automatic adoption of existing clusters (we will provide instructions to allow users to opt-in for existing cluster)
+* Management of anything that falls out of the scope of Machine API
+
+## Proposal
+
+A new CRD `ControlPlaneMachineSet` will be introduced into OpenShift and a respective `control-plane-machine-set-
+operator` will be introduced as a new second level operator within OpenShift to perform the operations described in
+this proposal.
+
+The new CRD will define the specification for managing (creating, adopting, updating) Machines for the Control Plane.
+The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
+cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
+of updated configuration to the Control Plane Machines.
+
+### User Stories
+
+- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
+  control plane as and when the demand on the control plane changes
+- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
```

Comment:
Ahh right, this is in terms of a `Failed` phase machine, so normally this would be that the instance has been removed from the cloud provider for some reason, or possibly a MachineHealthCheck has determined that it's unhealthy and should be replaced, I will clarify